vermouth/processors/do_links.py

```diff
@@ -261,36 +262,147 @@ def _build_link_interaction_from(molecule, interaction, match):
+
+                if interaction_key_rev in self.applied_links[inter_type]:
+                    del self.applied_links[inter_type][interaction_key]
+        return molecule
+
+    def add_edges(self, molecule, match):
+        """
+        Add edges from link to the molecule.
+
+        Parameters
+        ----------
+        molecule: :class:`vermouth.Molecule`
+        match: dict
+            dict relating nodes in `self.current_link` to those in molecule
+        """
+        # now we already add the edges of this link
+        # links can overwrite each other but the edges must be the same
+        # this is safer than using the make_edge method because it accounts
+        # for edges written in the edges directive
+        for edge in self.current_link.edges:
+            molecule.add_edge(match[edge[0]], match[edge[1]])
+        return molecule
+
     def run_molecule(self, molecule):
-        # TODO: Separate this into a function
         links = molecule.force_field.links
         _nodes_to_remove = []
+        # loop over all links in force-field
         for link in links:
+            self.current_link = link
+            # match link based on graph isomorphism and link-order
             matches = match_link(molecule, link)
+            # for each match replace attributes, interactions and edges
             for match in matches:
-                for node, node_attrs in link.nodes.items():
-                    if 'replace' in node_attrs:
-                        if node_attrs['replace'].get('atomname', False) is None:
-                            _nodes_to_remove.append(match[node])
-                        else:
-                            node_mol = molecule.nodes[match[node]]
-                            node_mol.update(node_attrs['replace'])
-                for inter_type, interactions in link.removed_interactions.items():
-                    for interaction in interactions:
-                        interaction = _build_link_interaction_from(molecule, interaction, match)
-                        try:
-                            molecule.remove_matching_interaction(inter_type, interaction)
-                        except ValueError:
-                            pass
-                for inter_type, interactions in link.interactions.items():
-                    for interaction in interactions:
-                        interaction = _build_link_interaction_from(molecule, interaction, match)
-                        molecule.add_or_replace_interaction(inter_type, *interaction, link.citations)
-
-            molecule.remove_nodes_from(_nodes_to_remove)
+                self.replace_attributes(molecule, match)
+                self.add_interactions(molecule, match)
+                self.remove_interactions(molecule, match)
+                self.add_edges(molecule, match)
+
+            # we remove the nodes scheduled for removal
+            molecule.remove_nodes_from(self.nodes_to_remove)
```

Comment:
Add interactions first, then remove nodes. That will be slower, but it could be link 1 adds an interaction on a node that link 2 removes.